### PR TITLE
James 2083 migration

### PR DESCRIPTION
**Pull request by strokyl**
WIP:
 - test should be more exhaustive and check that Attachment are properly migrated
 - we have to discuss my RawMessageWithoutAttachment. The issue is that before my commit most of our function expected message with their Mailbox context but one we don't need it, two we can't getting it just by scanning the table to migrate.
I change only method that I needed for my migration but maybe we can change more method. Also if we keep RawMessageWithoutAttachment IMO we should make MessageWithoutAttachment use it by composition instead of reusing all field.

### CODE REVIEW

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/AttachmentLoader.java [hunk 1]
```diff
@@ -63,6 +64,22 @@ public AttachmentLoader(CassandraAttachmentMapper attachmentMapper) {
         }
     }
 
+    public CompletableFuture<Stream<SimpleMessage>> addAttachmentToMessages(Stream<Pair<RawMessageWithoutAttachment,
```
Comment (strokyl) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/AttachmentLoader.java]: Here I introduce a big code duplication between the two addAttachmentToMessages IMO the first one should call the second one.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageDAO.java [hunk 2]
```diff
@@ -212,21 +235,29 @@ private ByteBuffer toByteBuffer(InputStream stream) throws IOException {
...
-    private Pair<MessageWithoutAttachment, Stream<MessageAttachmentRepresentation>> message(Row row, ComposedMessageIdWithMetaData messageIdWithMetaData, FetchType fetchType) {
-        ComposedMessageId messageId = messageIdWithMetaData.getComposedMessageId();
+    private Pair<MessageWithoutAttachment, Stream<MessageAttachmentRepresentation>> message(
+        Row row,
+        FetchType fetchType,
+        ComposedMessageIdWithMetaData composedMessageIdWithMetaData
+    ) {
+        Pair<RawMessageWithoutAttachment, Stream<MessageAttachmentRepresentation>> rawPair = rawMessage(row, fetchType);
+
+        return Pair.of(
```
Comment (strokyl) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageDAO.java]: I did something like that twice in this PR. IMO it could be great to add a PairUtil in the same package than OptionalConvertor. This pair util will have two statics methods mapLeft and mapRight.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/MessageWithoutAttachment.java [hunk 3]
```diff
@@ -59,6 +59,10 @@ public MessageWithoutAttachment(MessageId messageId, Date internalDate, Long siz
         this.modSeq = modSeq;
     }
 
+    public RawMessageWithoutAttachment toRawMessageWithoutAttachment() {
```
Comment (strokyl) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/MessageWithoutAttachment.java]: Maybe it would be better to make a interface for message without attachment as we did for Message and SimpleMessage that have attachment.
This will allow to use all method that could take a RawMessageWithoutAttachment (IE a mail with it's mailbox context) take care of message with the context for free.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/migration/MigrationTracking.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        logProcess();
+
+        return new State(numOfSuccess.get(), error);
+    }
+
+    public State getState() {
+        return new State(numOfSuccess.get(), numOfError.get());
+    }
+
+    private void logProcess() {
```
Comment (strokyl) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/migration/MigrationTracking.java]: Should I be more generic and let guice inject a module to customize the tracking process ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/migration/V1ToV2Migration.java [hunk 5]
```diff
@@ -85,17 +88,26 @@ public void stop() {
...
-            .thenApply(this::submitMigration);
+            .thenApply(message -> {
+                this.submitMigration(Pair.of(message.getLeft().toRawMessageWithoutAttachment(), message.getRight()));
+
+                return message;
+            });
     }
 
-    private Pair<MessageWithoutAttachment, Stream<MessageAttachmentRepresentation>> submitMigration(Pair<MessageWithoutAttachment, Stream<MessageAttachmentRepresentation>> messageV1) {
+    private void submitMigration(
```
Comment (strokyl) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/migration/V1ToV2Migration.java]: Those three line should be in only one line

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/CassandraMailboxModule.java [hunk 6]
```diff
@@ -33,7 +33,6 @@
 import org.apache.james.mailbox.cassandra.CassandraSubscriptionManager;
 import org.apache.james.mailbox.cassandra.ids.CassandraId;
 import org.apache.james.mailbox.cassandra.ids.CassandraMessageId;
-import org.apache.james.mailbox.cassandra.mail.CassandraMailboxDAO;
```
Comment (strokyl) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/CassandraMailboxModule.java]: This should be in a separate commit "Clean import"

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/migration/MigrationTracking.java (hunk 4 above)
Comment (aduprat) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/migration/MigrationTracking.java]: No, that's fine for now.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/table/CassandraMessageV1Table.java [hunk 7]
```diff
@@ -31,7 +31,7 @@
     String TEXTUAL_LINE_COUNT = "textualLineCount";
     String BODY_CONTENT = "bodyContent";
     String HEADER_CONTENT = "headerContent";
-    String PROPERTIES = "properties";
+    String PROPERTIES = "propertyBuilder";
```
Comment (strokyl) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/table/CassandraMessageV1Table.java]: error

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/table/CassandraMessageV2Table.java [hunk 8]
```diff
@@ -31,7 +31,7 @@
     String TEXTUAL_LINE_COUNT = "textualLineCount";
     String BODY_CONTENT = "bodyContent";
     String HEADER_CONTENT = "headerContent";
-    String PROPERTIES = "properties";
+    String PROPERTIES = "propertyBuilder";
```
Comment (strokyl) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/table/CassandraMessageV2Table.java]: error

File: mailbox/jcr/src/main/java/org/apache/james/mailbox/jcr/mail/model/JCRMailboxMessage.java [hunk 9]
```diff
@@ -309,13 +309,13 @@ public void merge(Node node) throws RepositoryException, IOException {
             for (Property prop : currentProperties) {
                 newProperties.add(new JCRProperty(prop, logger));
             }
-            // remove old properties, we will add a bunch of new ones
+            // remove old propertyBuilder, we will add a bunch of new ones
```
Comment (strokyl) [inline mailbox/jcr/src/main/java/org/apache/james/mailbox/jcr/mail/model/JCRMailboxMessage.java]: error

File: mailbox/hbase/src/main/java/org/apache/james/mailbox/hbase/mail/HBaseMailboxMessage.java [hunk 10]
```diff
@@ -208,8 +208,8 @@ public long getBodyOctets() {
     }
 
     /**
-     * Gets a read-only list of meta-data properties.
-     * For properties with multiple values, this list will contain
+     * Gets a read-only list of meta-data propertyBuilder.
+     * For propertyBuilder with multiple values, this list will contain
```
Comment (strokyl) [inline mailbox/hbase/src/main/java/org/apache/james/mailbox/hbase/mail/HBaseMailboxMessage.java]: error

File: mailbox/hbase/src/main/java/org/apache/james/mailbox/hbase/mail/HBaseMailboxMessage.java [hunk 11]
```diff
@@ -99,7 +99,7 @@
     
     /**
      * Create a copy of the given message.
-     * All properties are cloned except mailbox and UID.
+     * All propertyBuilder are cloned except mailbox and UID.
```
Comment (strokyl) [inline mailbox/hbase/src/main/java/org/apache/james/mailbox/hbase/mail/HBaseMailboxMessage.java]: error

File: mailbox/jcr/src/main/java/org/apache/james/mailbox/jcr/mail/model/JCRMailboxMessage.java [hunk 12]
```diff
@@ -309,13 +309,13 @@ public void merge(Node node) throws RepositoryException, IOException {
...
             NodeIterator iterator = node.getNodes("messageProperty");
             while (iterator.hasNext()) {
                 iterator.nextNode().remove();
             }
 
-            // store new properties
+            // store new propertyBuilder
```
Comment (strokyl) [inline mailbox/jcr/src/main/java/org/apache/james/mailbox/jcr/mail/model/JCRMailboxMessage.java]: error

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/JPAProperty.java [hunk 13]
```diff
@@ -41,7 +41,7 @@
     @Index(name = "INDEX_PROPERTY_MSG_ID", columnNames = { "MAILBOX_ID", "MAIL_UID" })
     private long id;
 
-    /** Order within the list of properties */
+    /** Order within the list of propertyBuilder */
```
Comment (strokyl) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/JPAProperty.java]: error

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java [hunk 14]
```diff
@@ -259,7 +259,7 @@ public AbstractJPAMailboxMessage(JPAMailbox mailbox, Date internalDate, Flags fl
     }
 
     /**
-     * Constructs a copy of the given message. All properties are cloned except
+     * Constructs a copy of the given message. All propertyBuilder are cloned except
```
Comment (strokyl) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java]: error

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java [hunk 15]
```diff
@@ -348,7 +348,7 @@ public String getSubType() {
     }
 
     /**
-     * Gets a read-only list of meta-data properties. For properties with
+     * Gets a read-only list of meta-data propertyBuilder. For propertyBuilder with
```
Comment (strokyl) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java]: error

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java [hunk 16]
```diff
@@ -183,7 +184,19 @@ public void setInMailboxes(MessageId messageId, List<MailboxId> mailboxIds, Mail
...
+        return MessageUtil.buildMailboxMessage()
+            .messageId(messageIdFactory.generate())
+            .internalDate(internalDate)
+            .size(size)
+            .bodyStartOctet(bodyStartOctet)
+            .content(content)
+            .flags(flags)
+            .propertyBuilder(propertyBuilder)
+            .modSeq(0)
+            .uid(null)
```
Comment (strokyl) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java]: uid(null) is going to fal

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 17]
```diff
@@ -475,7 +476,30 @@ private boolean hasNonInlinedAttachment(List<MessageAttachment> attachments) {
...
+            .messageId(messageIdFactory.generate())
+            .internalDate(internalDate)
+            .size(size)
+            .bodyStartOctet(bodyStartOctet)
+            .content(content)
+            .flags(flags)
+            .propertyBuilder(propertyBuilder)
+            .mailboxId(getMailboxEntity().getMailboxId())
+            .attachments(attachments)
+            .uid(null)
```
Comment (strokyl) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: this is going to fail

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/StandardNames.java [hunk 18]
```diff
@@ -19,11 +19,11 @@
 package org.apache.james.mailbox.store.mail.model;
 
 /**
- * <p>Conventional meta-data names for standard properties.</p>
+ * <p>Conventional meta-data names for standard propertyBuilder.</p>
  * <h4>Conventions</h4>
  * <p>
  * The domain containing the namespace should be controlled by the definer.
- * So, all properties defined here begin with <code>http://james.apache.org</code>.
+ * So, all propertyBuilder defined here begin with <code>http://james.apache.org</code>.
```
Comment (strokyl) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/StandardNames.java]: error

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MailboxContextBuilder.java [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+        modSeq(idWithMetaData.getModSeq());
+        flags(idWithMetaData.getFlags());
+        return this;
+    }
+
+    public HasMailboxContext build() {
+        Preconditions.checkState(mailboxId.isPresent(), "mailboxId is required");
+
+        //Preconditions.checkState(uid.isPresent(), "uid is required");
+        //not sure if this is good
```
Comment (strokyl) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MailboxContextBuilder.java]: big question test create SimpleMailboxMessage without uid should we adapt test our new remplacement of SimpleMailboxMessage?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/PropertyBuilder.java [hunk 20]
```diff
@@ -53,7 +53,7 @@
 import com.google.common.base.Predicate;
 
 /**
- * Builds properties
```
Comment (strokyl) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/PropertyBuilder.java]: error

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/PropertyBuilder.java [hunk 21]
```diff
@@ -490,7 +490,7 @@ public String toString()
     {
         return "PropertyBuilder ( "
         + " textualLineCount = " + this.textualLineCount
-        + " properties = " + this.properties
+        + " propertyBuilder = " + this.properties
```
Comment (strokyl) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/PropertyBuilder.java]: error

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MailboxContextBuilder.java (hunk 19 above)
Comment (aduprat) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MailboxContextBuilder.java]: No, we must have several implementation depending on the mailbox implementation.
For example in JPA, the uid is given by the database; so we can't have an immutable MailboxMessage for this implementation...

**Comment by chibenwa** [comment]
Ok, Attachment presence here is defined **at runtime** using the fetchType.

FetchType.FULL -> Attachment will be here

FetchType.Metadata -> No attachment

Keeping this as a runtime information then seems legitimate to me.

**Comment by chibenwa** [comment]
As discussed during the daily, we don't have enough time to carry out this refactoring
